### ISSUE TITLE
Contouring helpers

### ISSUE THREAD

**Issue by aph42** [pygeode/pygeode#94]
For the past year or two I've been using a set of helper functions to define
sets of keyword arguments to pass to showvar to set up contouring in sensible
ways.  Recently I've been working on incorporating them into the interface for
showvar itself, as an attempt to make a cleaner interface and to set up more
powerful defaults.  This code is now in the plt_cnthelper branch. It does
affect the default contouring used by showvar when no arguments are provided,
but should not affect any code which sets specific arguments, or line plots.

There is reasonably functional code in this branch - any comments are much welcome. 
I've described the options below and given a bunch of examples.

The keyword logic is handled by plot/pyg_helpers.py:_parse_autofmt_kwargs(). It
recognizes five different types of plots, which can be set explicitly with
the `type' keyword:
- raw: Arguments are passed on directly to vcontour, as before. If 'clines' or 'clevs'
  are set explicitly, this type is selected.
- clf: Filled contours are used. This is now the default. Contour
  spacing is defined in terms of a set spacing ('cdelt'), which defines the
  major divisions of the colorbar: there are 'ndiv' divisions which each consist
  of 'nf' filled contours and 'nl' contour lines (the latter can be zero, the
  former cannot). Each division is marked with a tick on the colorbar.
  
  There are two styles of colormaps: divergent, which are symmetric and have an
  equal number of divisions above and below a mid point (usually zero), and
  sequential, which just have a set number of divisions (usually not centred
  around zero). There are default colormaps defined for divergent colormaps with
  1, 2, 3, or 4 divisions, and for sequential colormaps with 1 through 6
  divisions.
  
  If no arguments are given, the contour intervals and colormap are determined
  automatically by the function plot/cnt_helpers.py:guessclimits(). This can
  choose divergent vs. sequential based on whether the data looks like it's
  close to being centred around zero or not, and chooses an appropriate (and
  round) spacing 'cdelt' such that the majority of the data is within the
  bounds, unless the distribution of data is highly skewed (this can probably
  use some tweaking). At present 3 divisions are used for divergent data, and 5
  for sequential data, though both the style and number of divisions can be set
  explicitly with the keywords 'style' (set to 'div' or 'seq') and 'ndiv'.
  
  If the spacing 'cdelt' is set explicitly, divergent data is assumed unless the
  'min' argument is also provided, which sets the lower bound of the contours.
  
  The colormaps themselves are defined by a set of values in plot/cm.py
  (hsv_div, hsv_seq) which determine the hue for each division, as well as the
  bounds of the saturation and value. I have thought about putting these in the
  pygrc file somewhere or otherwise making these more easily customizable but
  haven't figured out the right way to do that yet.
  
  See also cnt_helpers.py:clfdict()
  
  A couple of samples: (after 'import pygeode as pyg; from pygeode.tutorial import t1')

```
 pyg.showvar(t1.Temp)            # sequential
 pyg.showvar(t1.Temp, ndiv=3)    # sequential, 3 divisions
 pyg.showvar(t1.Temp, ndiv=2, nf=9, nl=3)    # sequential, 2 divisions, 9 filled & 3 lines per division
 pyg.showvar(t1.Temp - 273.15)      # divergent
 pyg.showvar(t1.Temp - 273.15, ndiv=2)      # divergent, 2 divisions
 pyg.showvar(t1.Temp - 273.15, cdelt=8, ndiv=4)    # divergent, explicit spacing
 pyg.showvar(t1.Temp - 273.15, cdelt=10, min=-20, nf=5, nl=1)    # sequential, explicit spacing; 5 filled
```
- cl: Unfilled contour plot. Contours are defined in terms of an interval ('cdelt'); in this case
  'ndiv' sets the number of contours to use. If `ndiv' is specified alone, a
  contour interval is chosen to span the values of the data set (again, up to
  large skew). By default this is set to 10. If cdelt is set, the contours are
  by default centred around zero; alternativley the mid value or minimum can be
  specified with 'mid' or 'min', respectively. Finally, the zero contour can be omitted
  by setting 'nozero = True'.
  
  See also cnt_helpers.py:cldict()
  
  Some examples:

```
 pyg.showvar(t1.Temp, type='cl')
 pyg.showvar(t1.Temp, type='cl', ndiv=15)
 pyg.showvar(t1.Temp, type='cl', cdelt=5, mid=270)
 pyg.showvar(t1.Temp - 273.15, type='cl', cdelt=5, nozero=True)
```
- log or log1s: Filled contours, spaced logarithmically, useful for variabes
  with large dynamic ranges. In this case the logarithmically spaced contours
  are specified by a minimum contour level ('cmin'), a number of divisions
  ('ndiv,' as in the clf type), and the logarithmic base ('cdelt'; each division
  is one power of this base). The minimum contour level 'cmin' must be
  explicitly given at present. The number of divisions is set by default to 5,
  and the base is set to 10. The default colourmap is the same as for the
  sequential data style.
  
  See also cnt_helpers.py:log1sdict()
  
  Some examples:

```
 pyg.showvar(pyg.exp((t1.Temp/100.)**3), type='log1s', cmin=1e7)
 pyg.showvar(pyg.exp((t1.Temp/100.)**3), type='log1s', cmin=1e3, cdelt=1e2)
```
- log2s: Filled contours, spaced logarithmically on either side of 0, with a
  region of linear scaling to connect the two. Here 'cmin' specifies the
  transition point between linear and logarithmic spacing.  The first division
  either side of zero is linear, the subsequent ones are logarithmic; it doesn't
  make much sense to specify ndiv less than 2 in this case (though you can do
  so). As in the one sided case, 'cdelt' specifies the base factor that each
  logarithmic division represents. The default colourmap is the same as for the
  divergent data style.
  
  See also cnt_helpers.py:log2sdict()
  
  Some examples:

```
 t = pyg.exp((t1.Temp/100.)**2)*pyg.sind(t1.lon) # some two-sided high-dynamical range data
 pyg.showvar(t, type='log2s', cmin=1e3)
 pyg.showvar(t, type='log2s', cmin=1e1, cdelt=1e2, nf=4)
 pyg.showvar(t, type='log2s', cmin=1e3, cdelt=1e1, nozero=True, ndiv=2)
```

**Comment by neishm** [comment]
I tried out some of the examples you posted, and it looks good to me.  I like the colour map a lot better than the current default.  However, my only concern is that the change in default colour map may cause confusion for anybody using this method in their scripts.  It may not be immediately apparent how to revert to the old colour map.  After some fiddling, I was able to emulate the old version with parameters `cmap=pl.cm.jet, nl=0`.

Maybe once this code is merged, we could have a warning/information message if someone is using `showvar` with no explicit cmap, letting them know the default has changed and how to revert it if they want?

Alternatively, we could just change the default and put the relevant info in the documentation, so if someone is vexed enough by the change they can find out how to deal with it themselves.

Actually, come to think of it matplotlib is changing their default colormap in 2.0 so people will have to deal with changes anyway, so it's probably not a big deal if we change things too.

If the functionality is complete, I'm okay with it being merged into the master branch before release-1.1.
